nft_ingester/src/program_transformers/token/mod.rs

```diff
@@ -136,21 +147,16 @@ pub async fn handle_token_program_account<'a, 'b, 'c>(
                 )
                 .one(db)
                 .await?;
-            if let Some(asset) = asset_update {
-                let mut active: asset::ActiveModel = asset.clone().into();
-                active.supply = Set(m.supply as i64);
-                active.supply_mint = Set(Some(key_bytes));
-
-                // Update owner_type based on the supply.
-                if asset.owner_type == OwnerType::Unknown {
-                    active.owner_type = match m.supply.cmp(&1) {
-                        std::cmp::Ordering::Equal => Set(OwnerType::Single),
-                        std::cmp::Ordering::Greater => Set(OwnerType::Token),
-                        _ => NotSet,
-                    }
-                }
-
-                active.save(db).await?;
+            if let Some(_asset) = asset_update {
+                upsert_assets_mint_account_columns(
+                    AssetMintAccountColumns {
+                        mint: key_bytes,
+                        supply: m.supply,
+                        slot_updated_mint_account: account_update.slot(),
+                    },
+                    db,
+                )
+                .await?;
```

Comment:
We've recently done all these studies for race conditions due to concurrent processes, one pattern I keep an eye out for and flag as potential issue is the one where there's a separate read and write SQL statements, which is happening here.

Consider this scenario:

First process: we are indexing a token with supply of 1, and one record is returned by the `find_by_id` (`OwnerType::Unknown` and `Column::Supply.eq(1)`),
Second (concurrent) process indexing after token is burned, sets supply to 0.
First process would erroneously set token back to supply of 1 with the `upsert_assets_mint_account_columns`

Should we combine the SQL statements affecting this asset row?  Or use a transaction with an explicit row lock?

Reply:
Yeah you are right, its protected exactly by the new slot updated value, as the change in token supply would be processed with this same call in the other process.  For some reason I was thinking it was done elsewhere.